packages/txservice/src/dispatch.ts

```diff
@@ -0,0 +1,195 @@
+import { BigNumber } from "ethers";
+import { BaseLogger } from "pino";
+import PriorityQueue from "p-queue";
+
+import { ChainRpcProvider } from "./provider";
+import { GasPrice, WriteTransaction } from "./types";
+import { Transaction } from "./transaction";
+import { TimeoutError, TransactionReverted } from "./error";
+
+// TODO: sep file
+// Thread safe stack manager.
+class TransactionBuffer {
+  // Both of the following are functionally treated as stacks (LIFO).
+  private validated: Transaction[] = [];
+  private pending: Transaction[] = [];
+  // Should be used to access either of the stacks.
+  private readonly accessQueue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  // Push a new tx to pending.
+  public async push(transaction: Transaction) {
+    // TODO: We may want to cap the pending stack here.
+    await this.accessQueue.add(() => {
+      this.pending.push(transaction);
+    });
+  }
+
+  // Move latest pending to completed.
+  // @returns boolean whether operation was successful, or if there was no pending tx to shift.
+  public async shift(id: string): Promise<boolean> {
+    // TODO: Trim the validated stack when it gets "too long" to save memory space.
+    return await this.accessQueue.add((): boolean => {
+      const transaction = this.pending.find((transaction) => transaction.id === id);
+      if (transaction == null) {
+        return false;
+      }
+      this.validated.push(transaction);
+      return true;
+    });
+  }
+
+  /**
+   * Get the current nonce value.
+   *
+   * @remarks
+   * Caller should still be prepared to get the incorrect nonce back. For instance,
+   * if the provider that just handled our sent tx has suddenly gone offline, this
+   * method may give the wrong nonce. This can be solved by making additional calls to
+   * submit the tx.
+   *
+   * @returns A number value for the current nonce.
```

Comment:
`current` nonce is not specific enough -- do you mean the current nonce that a tx should be submitted at or the current nonce that needs to be cleared

Reply:
Seems like (as paired with below) its always the nonce that you will submit a transaction at

Reply:
Fuck you right. Should be more specific. We could make this `getLastNonce` or we could just send back `(await this.getLastTx())?.nonce + 1;`